rewardbench/utils.py

```diff
@@ -407,6 +418,161 @@ def filter_long_turns(batch):
     return dataset, subsets
 
 
+def reroll_and_score_dataset(dataset, best_of, cols_to_combine=["text", "scores"]):
+    # Convert to pandas DataFrame for easier manipulation
+    df = dataset.to_pandas()
+
+    # Get the number of groups
+    n_groups = len(df) // best_of
+    if len(df) % best_of != 0:
+        raise ValueError(f"Dataset length ({len(df)}) is not divisible by best_of ({best_of})")
+
+    rerolled_rows = []
+
+    # Process each group of best_of rows
+    for i in range(n_groups):
+        start_idx = i * best_of
+        group = df.iloc[start_idx : start_idx + best_of]
+
+        # Create new row
+        new_row = {}
+
+        # Handle text and score columns - combine into lists
+        for col in cols_to_combine:
+            new_row[col] = group[col].tolist()
+
+        # Result is 1 if the first entry (chosen) is scored the highest, 0 otherwise
+        new_row["results"] = 1 if np.argmax(new_row["scores"]) == 0 else 0
+
+        # Handle all other columns - verify they're identical and take first value
+        other_columns = [col for col in df.columns if col not in cols_to_combine]
+        for col in other_columns:
+            values = group[col].unique()
+            if len(values) != 1:
+                raise ValueError(f"Column {col} has different values within group {i}: {values}")
+            new_row[col] = values[0]
+
+        rerolled_rows.append(new_row)
+
+    # Create new dataset
+    rerolled_df = pd.DataFrame(rerolled_rows)
+    rerolled_dataset = Dataset.from_pandas(rerolled_df)
+
+    return rerolled_dataset
+
+
+def load_bon_dataset_v2(
+    dataset: str,
+    custom_dialogue_formatting: bool = False,
+    conv: Conversation = None,
+    tokenizer: PreTrainedTokenizer = None,
+    logger: logging.Logger = None,
+    keep_columns: List[str] = ["text_chosen", "text_rejected", "text", "id"],
+    local: bool = False,
+):
+    """
+    Loads the BON candidates dataset.
+    """
+
+    # load the data. Data can be a HuggingFace dataset or a local JSONL file
+    if not dataset:
+        raw_dataset = load_dataset(CORE_EVAL_SET_V2, split="train")
+
+    else:
+        if ".jsonl" in dataset:
+            raw_dataset = pd.read_json(dataset, lines=True)
+            raw_dataset = Dataset.from_pandas(raw_dataset)
+        elif local:
+            raw_dataset = load_from_disk(dataset)
+        else:
+            # change split if renamed
+            raw_dataset = load_dataset(dataset, split="train")
```

Comment:
Suggestion - get in the habit of renaming the splits in eval datasets away from train, so people don't think the eval dataset is a training dataset ;). Took releasing v1 to realize this.